Client/mods/deathmatch/logic/luadefs/CLuaDrawingDefs.cpp

```diff
@@ -566,6 +568,98 @@ int CLuaDrawingDefs::DxDrawImageSection(lua_State* luaVM)
     return 1;
 }
 
+int CLuaDrawingDefs::DxDrawPrimitive3D(lua_State* luaVM)
+{
+    // bool DxDrawPrimitive3D (string primitiveType, bool postGUI, table vertice1, ...)
+    D3DPRIMITIVETYPE    ePrimitiveType;
+    auto                pVecVertices = new std::vector<PrimitiveVertice>();
+    bool                bPostGUI;
+    CScriptArgReader argStream(luaVM);
+    argStream.ReadEnumString(ePrimitiveType);
+    argStream.ReadBool(bPostGUI, false);
+
+    while (argStream.NextIsTable())
+    {
+        std::vector<float> vecTableContent;
+        argStream.ReadNumberTable(vecTableContent);
+        switch (vecTableContent.size())
+        {
+            case Primitive3DVerticeSizes::VERT_XYZ:
+                pVecVertices->push_back(PrimitiveVertice{vecTableContent[0], vecTableContent[1], vecTableContent[2], (DWORD)0xFFFFFFFF});
+                break;
+            case Primitive3DVerticeSizes::VERT_XYZ_COLOR:
+                pVecVertices->push_back(PrimitiveVertice{vecTableContent[0], vecTableContent[1], vecTableContent[2], static_cast<DWORD>(vecTableContent[3])});
+                break;
+        }
+    }
+
+    if (!argStream.HasErrors())
+    {
+        if (g_pCore->GetGraphics()->IsValidPrimitiveSize(pVecVertices->size(), ePrimitiveType))
+        {
+            g_pCore->GetGraphics()->DrawPrimitive3DQueued(pVecVertices, ePrimitiveType, bPostGUI);
+            lua_pushboolean(luaVM, true);
+            return 1;
+        }       
+    }
+    else
+    {
+        m_pScriptDebugging->LogCustom(luaVM, argStream.GetFullErrorMessage());
+    }
+
+    // Failed
+    delete pVecVertices;
+    lua_pushboolean(luaVM, false);
+    return 1;
+}
+
+int CLuaDrawingDefs::DxDrawMaterialPrimitive3D(lua_State* luaVM)
+{
+    // bool DxDrawMaterialPrimitive3D (string primitiveType, dxMaterial material, bool postGUI, table vertice1, ...)
+    D3DPRIMITIVETYPE    ePrimitiveType;
+    auto                pVecVertices = new std::vector<PrimitiveMaterialVertice>();
+    CClientMaterial*    pMaterialElement;
+    bool                bPostGUI;
+    CScriptArgReader argStream(luaVM);
+    argStream.ReadEnumString(ePrimitiveType);
+    MixedReadMaterialString(argStream, pMaterialElement);
+    argStream.ReadBool(bPostGUI, false);
+
+    while (argStream.NextIsTable())
+    {
+        std::vector<float> vecTableContent;
```

Comment:
This **std::vector** is going to heap allocate memory for every single vertice. Move it out of the **while** loop and call `std::vector<T>::clear()` here. Don't worry, `.size()` will have the expected value.

Reply:
@CrosRoad95 It seems `vecTableContent.clear()` is still missing in the latest version?

Reply:
@CrosRoad95

> and call std::vector::clear() here

seems to be missing or reason why not?